interop-future/jvm/src/main/scala/scalaz/zio/interop/Future.scala

```diff
@@ -0,0 +1,199 @@
+      unsafeRun(ec, IO.syncThrowable(body).fork)
+
+    final def sequence[A](in: List[Future[A]])(implicit ec: ExecutionContext): Future[List[A]] =
+      unsafeRun(ec, IO.collectAll(in.map(_.join)).fork)
+
+    final def sequence[A](in: Vector[Future[A]])(implicit ec: ExecutionContext): Future[Vector[A]] =
+      unsafeRun(ec, IO.collectAll(in.map(_.join)).map(_.toVector).fork)
+
+    final def sequence[A](in: Seq[Future[A]])(implicit ec: ExecutionContext): Future[Seq[A]] =
+      unsafeRun(ec, IO.collectAll(in.map(_.join)).map(_.toSeq).fork)
+
+    final def firstCompletedOf[T](futures: Iterable[Future[T]])(implicit ec: ExecutionContext): Future[T] =
+      unsafeRun(ec, IO.absolve(IO.raceAll(IO.interrupt, futures.map(_.join.attempt))).fork)
+
+    final def find[T](futures: Iterable[Future[T]])(p: T => Boolean)(implicit ec: ExecutionContext): Future[Option[T]] =
+      unsafeRun(ec, (futures.foldLeft[IO[Throwable, Option[T]]](IO.interrupt) {
+        case (acc, future) => 
+          acc orElse (future.join.flatMap(t => if (p(t)) IO.succeed(t) else IO.interrupt).map(Some(_)))
+      } orElse IO.succeed(None)).fork)
+
+    final def foldLeft[T, R](futures: Iterable[Future[T]])(zero: R)(op: (R, T) => R)(implicit ec: ExecutionContext): Future[R] =
+      unsafeRun(ec, futures.foldLeft[IO[Throwable, R]](IO.succeed(zero)) {
+        case (acc, future) =>
+          acc.flatMap(r => future.join.map(op(r, _)))
+      }.fork)
+
+    final def fold[T, R](futures: Iterable[Future[T]])(zero: R)(op: (R, T) => R)(implicit ec: ExecutionContext): Future[R] =
+      foldLeft(futures)(zero)(op)
+
+    final def reduce[T, R >: T](futures: Iterable[Future[T]])(op: (R, T) => R)(implicit ec: ExecutionContext): Future[R] = 
+      futures.headOption match {
+        case None => Fiber.interrupt 
+        case Some(t) => 
+          val ts = futures.tail
+
+          unsafeRun(ec, t.join.map(t => fold[T, R](ts)(t)(op)))
+      }
+
+    final def reduceLeft[T, R >: T](futures: Iterable[Future[T]])(op: (R, T) => R)(implicit ec: ExecutionContext): Future[R] =
+      reduce[T, R](futures)(op)
+
+    final def traverse[A, B](in: List[A])(fn: A => Future[B])(implicit ec: ExecutionContext): Future[List[B]] =
+      unsafeRun(ec, IO.foreach(in)(a => fn(a).join).fork)
+
+    final def traverse[A, B](in: Vector[A])(fn: A => Future[B])(implicit ec: ExecutionContext): Future[Vector[B]] =
+      unsafeRun(ec, IO.foreach(in)(a => fn(a).join).map(_.toVector).fork)
+
+    final def traverse[A, B](in: Seq[A])(fn: A => Future[B])(implicit ec: ExecutionContext): Future[Seq[B]] =
+      unsafeRun(ec, IO.foreach(in)(a => fn(a).join).map(_.toSeq).fork)
+  }
+
+  implicit class FutureSyntax[T](val value: Future[T]) extends AnyVal {
+    final def onSuccess[U](pf: PartialFunction[T, U])(implicit ec: ExecutionContext): Unit = 
+      unsafeRun(ec, value.join.flatMap(t => if (pf.isDefinedAt(t)) IO.syncThrowable(pf(t)) *> IO.unit else IO.unit).fork.void)
```

Comment:
or
```scala
(t => IO.syncThrowable(pf(t)).void.when(pf.isDefinedAt(t)))
```
😇

Reply:
`isDefinedAt` is not safe in the presence of side-effects when PartialFunction is created with `Function.unlift`, and the only reason you'd call `onSuccess` is for side-effects.

Could be
```scala
(t => IO.syncThrowable(pf.andThen(_ => ()).applyOrElse(t, _ => ())))
```

Reply:
Fixed it by using `pf.lift` and then always applying.

You're right we probably have to assume `pf.isDefinedAt` has side-effects. 😱 🔫 💥